### PR TITLE
Add Streams support

### PR DESCRIPTION
**Pull request by supercaracal**
I'd like to use Redis Streams.

* [Introduction to Redis Streams](https://redis.io/topics/streams-intro)
* [Redis Streams Commands](https://redis.io/commands#stream)

### CODE REVIEW

**Comment by supercaracal** [comment]
https://travis-ci.org/redis/redis-rb/jobs/458363796

> Redis::CommandError: IOERR error or timeout reading to target instance

It is related https://github.com/redis/redis-rb/pull/789

File: lib/redis.rb [hunk 1]
```diff
@@ -2948,6 +3268,43 @@ def method_missing(command, *args)
...
+      return {} if reply.nil?
+      reply.map do |stream_key, entries|
+        [stream_key, HashifyStreamEntries.call(entries)]
+      end.to_h
+    }
+
+  HashifyStreamEntries =
+    lambda { |reply|
+      reply.map do |entry_id, values|
+        [entry_id, Hash[values.each_slice(2).to_a]]
```
Comment (byroot) [inline lib/redis.rb]: `values.each_slice(2).to_h` is simpler.

File: lib/redis.rb [hunk 2]
```diff
@@ -2803,6 +2803,326 @@ def geodist(key, member1, member2, unit = 'm')
...
+  #   redis.xlen('mystream')
+  #
+  # @param key [String] the stream key
+  #
+  # @return [Integer] the number of entries
+  def xlen(key)
+    synchronize { |client| client.call([:xlen, key]) }
+  end
+
+  # Fetches entries from one or multiple streams. You're able to use blocking if needed.
```
Comment (byroot) [inline lib/redis.rb]: `You're able to use blocking if needed.` I think this needs to be phrased differently.

Maybe `Optionally blocking` ?

File: lib/redis.rb [hunk 3]
```diff
@@ -2803,6 +2803,326 @@ def geodist(key, member1, member2, unit = 'm')
...
+  # @param count [Integer] the number of entries as limit
+  #
+  # @return [Hash{String => Hash}] entries
+  def xrange(key, first: '-', last: '+', count: nil)
+    args = [:xrange, key, first, last]
+    args.concat(['COUNT', count]) unless count.nil?
+    synchronize { |client| client.call(args, &HashifyStreamEntries) }
+  end
+
+  # Fetches entries of the stream ordered by descending.
```
Comment (byroot) [inline lib/redis.rb]: `ordered by descending.` -> `in descending order`.

**Comment by supercaracal** [comment]
ref: https://github.com/andymccurdy/redis-py/pull/1040

**Comment by supercaracal** [comment]
I have fixed the following additional things.

* Fix several wrong or not enough YARD comments (e.g. `@return` tag)
* Add `#_xread` private method for DRY between `#xread` and `#xreadgroup`

**Comment by supercaracal** [comment]
I have fixed as the following.

* Refactor optional parameters of methods
* Change option naming concerns `#xtrim` and `#xadd` methods into a more meaningful word
  * `s/modifier/approximately/`

File: test/helper.rb [hunk 4]
```diff
@@ -159,6 +159,10 @@ def target_version(target)
       end
     end
 
+    def omit_version(min_ver)
```
Comment (byroot) [inline test/helper.rb]: This isn't used anywhere. Can you remove it from the PR please?

**Comment by byroot** [comment]
I think this looks very good. 

It just need a rebase and some minor ajustement.

I'd like to merge it and play with it a bit, and eventually release a 4.1.0 with this & the cluster support.

File: test/helper.rb (hunk 4 above)
Comment (supercaracal) [reply test/helper.rb]: The above helper method is used by the following linter.
https://github.com/redis/redis-rb/pull/799/files#diff-a14ac8847c163e36437c11a9aa3c7643R8-R11

Comment (byroot) [reply test/helper.rb]: Weird, that's a broken link for me and `ctrl + F omit_version` doesn't match anything.

Comment (byroot) [reply test/helper.rb]: Argh, nvm, GitHub chose to hide that part of the diff.

**Comment by byroot** [comment]
Can you rebase ? I'll merge afterward.

**Comment by supercaracal** [comment]
I have rebased. I appreciate for the your review.

File: test/helper.rb (hunk 4 above)
Comment (supercaracal) [reply test/helper.rb]: I'm sorry.

```ruby
def setup
  super
  omit_version(MIN_REDIS_VERSION)
end
```

**Pull request merged**